### PR TITLE
[WIP] TF Refactor Mark II

### PR DESCRIPTION
**Pull request by rsalmond**
See various README.md files for high level explanations of the moving parts.

As of right now `infra/bootstrap` and `infra/singletons` are correct. That is, there is a live TF state file in the s3 bucket which is up to date and `tf plan` does not want to change anything. There's a little bit left to do with `app/stage` and `app/prod` to bring them up to this same level, and nothing yet to do for `infra/prod` and `infra/stage` since we haven't created those things yet.

Once the state for each separate TF composition is up to date, we can merge this and remove the old original TF state file from the S3 bucket (which will then be superseded by the new split up ones)

### CODE REVIEW

File: infra/prod/tofu.tf [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    region         = "ca-central-1"
+    dynamodb_table = "terraform-state-locks"
+    encrypt        = true
+    progile        = "gpo"
+  }
+}
+
+provider "aws" {
+  region  = "ca-central-1"
+  progile = "gpo"
```
Comment (rsalmond) [inline infra/prod/tofu.tf:27]:
```suggestion
  profile = "gpo"
```

File: README.md [hunk 2]
```diff
@@ -32,6 +32,46 @@ https://docs.google.com/document/d/1-2-MSpd-g_i5UjxHVkZW8wpKUZORjLrzd8A-UhW2XAY/
...
+
+Contains staging infrastructure on which staging apps run.
+
+## [infra/prod](./infra/prod)
+
+Contains production infrastructure on which production apps run.
+
+## [app/stage](./app/stage)
+
+Containers resources which staging appls consume.
```
Comment (IanEdington) [inline README.md:57]:
```suggestion
Containers resources which staging apps consume.
```

File: infra/singletons/github_repos.tf [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  description   = "JavaScript (and TypeScript) client for CiviCRM API v4"
+  visibility    = "public"
+  has_downloads = true
+  has_projects  = true
+}
+
+resource "github_branch_default" "civicrm_api" {
+  repository = github_repository.civicrm_api.name
+  branch     = "main"
+}
```
Comment (IanEdington) [inline infra/singletons/github_repos.tf:106]: Some of these repos are archived. Is there value in keeping them under terraform?

File: infra/singletons/tofu.tf [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  }
+  /*
+  backend "s3" {
+    bucket         = "gpo-terraform-state"
+    key            = "infra/singletons/terraform.tfstate"
+    region         = "ca-central-1"
+    dynamodb_table = "terraform-state-locks"
+    encrypt        = true
+    profile        = "gpo"
+  }*/
```
Comment (IanEdington) [inline infra/singletons/tofu.tf:27]: Is this intentional? It looks like the other `tofu.tf` files have backends set up

File: .gitignore [hunk 5]
```diff
@@ -3,5 +3,5 @@
 *.tfstate
 *.tfstate.backup
 plan
-
+.zsh_config
```
Comment (IanEdington) [inline .gitignore:6]: aside: Is this for directory-level configuration loading/unloading? What tool do you use?

**Review by IanEdington — APPROVED** [review]
Looks awesome! Thanks for all your work on this!

File: infra/singletons/github_repos.tf (hunk 3 above)
Comment (rsalmond) [reply infra/singletons/github_repos.tf:106]: Probably not. I have only added the repos I was able to see before joining the github org. Added to #34

File: .gitignore (hunk 5 above)
Comment (rsalmond) [reply .gitignore:6]: It was a one off I tested for something that didn't work. I'll remove it :/

File: infra/singletons/tofu.tf (hunk 4 above)
Comment (rsalmond) [reply infra/singletons/tofu.tf:27]: ah yeah I was testing something, will revert

File: README.md (hunk 2 above)
Comment (rsalmond) [reply README.md:57]: Fixed locally.

**Comment by rsalmond** [comment]
Update: okay ALL the new separate sub-states / compositions / stacks / whatever you want to call them are up to date. That is, all the state has been sliced up out of the original state file and moved into thees separate state files. You can `tofu init && tofu plan` in any subdir containing a `tofu.tf` file (presence of this file indicates an isolated stack) and it will say "no changes".

For reason I cannot begin to understand the digital ocean spaces auth just _refuses_ to accept anything but env vars so for the time being I've added a helper script to run tf with spaces env vars until we can fix that. Note that I changed the key names in secrets.env to support this, but then decided against using `sops exec-env` because I didn't like having _every_ secret out when it wasn't necessary.

**Comment by rsalmond** [comment]
Okay I got rid of the `terraform init` errors that the pre-commit hooks were grumping about but I am not sure why the `tf validate` step is failing, cause they all look right.

```console
$ find . -name tofu.tf -exec bash -c 'cd "$(dirname {})" && tofu validate' \;
Success! The configuration is valid.
Success! The configuration is valid.
Success! The configuration is valid.
Success! The configuration is valid.
Success! The configuration is valid.
Success! The configuration is valid.
```

**Pull request merged**